django/contrib/auth/decorators.py

```diff
@@ -35,6 +35,27 @@ def _wrapper_view(request, *args, **kwargs):
 
             return redirect_to_login(path, resolved_login_url, redirect_field_name)
 
+        if asyncio.iscoroutinefunction(view_func):
+
+            @wraps(view_func)
+            async def _wrapper_view(request, *args, **kwargs):
+                if asyncio.iscoroutinefunction(test_func):
+                    result = await test_func(await request.auser())
+                else:
+                    result = test_func(await request.auser())
```

Comment:
Currently it's tested only with a `lambda`, it will not work when test cannot be called from async context, e.g.
```
        def my_test(user):
            return bool(Group.objects.filter(name=user.username).exists())
```

Reply:
Yes, you are right. it will raise an exception when `test_func` cannot be called from async context. But I think this is an expected behavior. That means, if developer provide a sync `test_func`, then it must can be called from async context. How do you think about?

Reply:
I agree with @felixxm, I think we should allow the user to pass a sync-only `test_func` to an `async` function using this decorator.

This makes the internal implementation more complicated, but it does allow users to have more flexibility in their implementations.

I think this is particularly important for a project that is "migrating" between sync and async, or using a mix of both contexts. We should not require them to implement 2 different test functions that do the same thing.